contracts/os/andromeda-kernel/src/execute.rs

```diff
@@ -918,251 +918,260 @@ pub fn unset_env(execute_ctx: ExecuteContext, variable: String) -> Result<Respon
-//                 self.message().message.clone(),
-//                 None,
-//             )
-//             .with_config(self.message().config.clone());
-
-//             let mut ctx = AMPCtx::new(info.sender.clone(), env.contract.address.clone(), 0, None);
-//             let username = ctx.try_add_origin_username(&deps.querier, &vfs_address);
-
-//             let hop = self.create_cross_chain_hop(
-//                 channel,
-//                 current_chain,
-//                 destination_chain,
-//                 ctx.get_origin(),
-//                 username.map(AndrAddr::from_string),
-//             );
-//             ctx.add_hop(hop);
-
-//             Ok(AMPPkt::new_with_ctx(ctx, vec![amp_msg]))
-//         }
-//         Some(mut amp_packet) => {
-//             // Update existing context
-//             let username = amp_packet
-//                 .ctx
-//                 .try_add_origin_username(&deps.querier, &vfs_address);
-
-//             let hop = self.create_cross_chain_hop(
-//                 channel,
-//                 current_chain,
-//                 destination_chain,
-//                 amp_packet.ctx.get_origin(),
-//                 username.map(AndrAddr::from_string),
-//             );
-//             amp_packet.ctx.add_hop(hop);
-
-//             // Remove chain reference from recipient
-//             amp_packet.messages[0].recipient =
-//                 AndrAddr::from_string(self.message().recipient.clone().get_raw_path().to_string());
-
-//             Ok(amp_packet)
-//         }
-//     }
-// }
+    //check if funds are present
+    if message.funds.is_empty() {
+        handle_ibc_direct(deps, info, env, ctx, message, channel_info)
+    } else {
+        handle_ibc_transfer_funds(deps, info, env, ctx, message, channel_info)
+    }
+}
 
-// fn handle_ibc_direct(
-//     deps: DepsMut,
-//     info: MessageInfo,
-//     env: Env,
-//     ctx: Option<AMPPkt>,
-//     sequence: u64,
-//     channel_info: ChannelInfo,
-// ) -> Result<Response, ContractError> {
-//     // Validate message is not empty
-//     ensure!(
-//         !Binary::default().eq(&self.message().message),
-//         ContractError::InvalidPacket {
-//             error: Some("Cannot send an empty message without funds via IBC".to_string())
-//         }
-//     );
+/// Builds an AMP packet for cross-chain communication via IBC
+///
+/// This function prepares a message to be sent across chains by:
+/// 1. Setting up the message content
+/// 2. Creating or updating the message context with routing information
+/// 3. Adding username data and hop information for cross-chain tracking
+fn build_ibc_context(
+    deps: &DepsMut,
+    info: &MessageInfo,
+    env: &Env,
+    channel: &str,
+    existing_packet: Option<AMPPkt>,
+    current_chain: String,
+    message: AMPMsg,
+    destination_chain: &String,
+) -> Result<AMPPkt, ContractError> {
+    // Get VFS address for username lookups
+    let vfs_address =
```

Comment:
I think it would be clearer if we did this directly before using it. So just before `// Add username information if available`